sdk-tools/src/sign.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright (c) 2023 The MobileCoin Foundation
+
+//! Builder wrapper around SgxSign.
+
+use std::{
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+/// This structure wraps the execution of sgx_sign commands.
+#[derive(Clone, Debug)]
+pub struct SgxSign {
+    /// The path to the sgx_sign executable.
+    sgx_sign_path: PathBuf,
+    /// Whether to ignore the presence of relocations in the enclave shared
+    /// object.
+    ignore_relocation_error: bool,
+    /// Whether to ignore .init sections in the enclave.
+    ignore_init_section_error: bool,
+    /// Whether to re-sign a previously signed enclave (default: false)
+    resign: bool,
+}
+
+impl Default for SgxSign {
+    /// Create a new SGX signing utility from the current environment.
+    fn default() -> Self {
+        let bin_dir = mc_sgx_core_build::sgx_bin_x64_dir();
+
+        SgxSign::from(bin_dir.join("sgx_sign"))
+    }
+}
+
+impl SgxSign {
+    /// Relocations are generally forbidden in the enclave shared object, this
+    /// tells the `sgx_sign` utility to ignore those errors.
+    #[must_use]
+    pub fn allow_relocations(mut self, allow: bool) -> Self {
+        self.ignore_relocation_error = allow;
+        self
+    }
+
+    /// Whether or not to allow .init sections in the enclave.
+    #[must_use]
+    pub fn allow_init_sections(mut self, allow: bool) -> Self {
+        self.ignore_init_section_error = allow;
+        self
+    }
+
+    /// Whether to re-sign a previously signed enclave (default: false)
+    #[must_use]
+    pub fn allow_resign(mut self, allow: bool) -> Self {
+        self.resign = allow;
+        self
+    }
+
+    /// Generate the command to sign the given enclave object with the given
+    /// private key and write the resulting enclave to the given path. Note
+    /// that online signatures are inherently insecure.
+    pub fn sign(
+        &mut self,
+        unsigned_enclave: &Path,
+        config_path: &Path,
+        private_key: &Path,
+        output_enclave: &Path,
+    ) -> Command {
```

Comment:
:wrench: I think IFF this suggestion is agreed on it should be a subsequent PR
Having this return a `Command` feels like it's leaking the abstraction. Clients will need to know what kind of output and return codes `sgx_sign` produces. I also think it's a bit misleading in that if I call a method named `sign` I expect it to sign something.
Suggest keeping the name but have this execute the `Command` and return something like a `Result<PathBuf>` with the signed enclave. I would move the current methods constant to a private method `build_sign_command()` so that it can be easily testable.

Reply:
IIRC, the reason I did it this way was because `Command` has three different execute methods: `spawn() -> Result<Child>`, `status() -> Result<ExitStatus>`, and `output() -> Result<Output>`, providing different levels of control.

So if you want to access stdout/stderr, it's capture in `output()`, if you only care about pass/fail, `status().success()` is less to worry about.

Reply:
Will be fixed in https://github.com/mobilecoinfoundation/sgx/issues/264